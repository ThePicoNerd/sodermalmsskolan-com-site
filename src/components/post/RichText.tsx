import Link from "next/link";
import React, { FunctionComponent, useMemo } from "react";
import ReactHtmlParser, { convertNodeToElement, Transform } from "react-html-parser";
import { breakpoints, media } from "../../styles/breakpoints";
import { fonts } from "../../styles/text";
import KatexText from "../katex/KatexText";
import InlineSkeleton from "../skeleton/InlineSkeleton";
import Skeleton from "../skeleton/Skeleton";
import Table from "../Table";

/**
 * Figure used in posts.
 *
 * @param {React.PropsWithChildren} props Standard HTML props.
 *
 * @returns {React.ReactElement} Rendered figure.
 */
const Figure: FunctionComponent = (props) => (
  <div
    css={{
      "--figure-margin-x": "0px",
      "--figure-margin-y": "4rem",

      margin: "var(--figure-margin-y) var(--figure-margin-x)",

      [media(breakpoints.small)]: {
        "--figure-margin-x": "-2rem",
      },

      [media(breakpoints.medium)]: {
        "--figure-margin-x": "-3rem",
      },
    }}
    {...props}
  />
);

/**
 * Function used to transform the HTML elements.
 *
 * @param {any} node The node.
 * @param {any} index Index of the element.
 *
 * @returns {React.ReactElement} The transformed element.
 */
const transform: Transform = (node, index) => {
  if (node.name === "a") {
    const { href } = node.attribs;

    return (
      <Link key={index} href={href} passHref>
        <a css={{
          color: "inherit",
          textDecoration: "underline",

          "&:hover": {
            color: "var(--link-color)",
          },
        }}
        >
          {node.children.map(transform)}
        </a>
      </Link>
    );
  }

  if (node.name === "figure") {
    return <Figure key={index}>{node.children.map(transform)}</Figure>;
  }

  if (node.name === "iframe") {
    return (
      <Figure key={index}>
        <div css={{
          position: "relative",
          display: "block",
          width: "100%",
          overflow: "hidden",

          "&::before": {
            paddingTop: "56.25rem",
            display: "block",
            content: "\"\"",
          },
        }}
        >
          {/* eslint-disable-next-line jsx-a11y/iframe-has-title */}
          <iframe
            css={{
              position: "absolute",
              top: 0,
              right: 0,
              bottom: 0,
              left: 0,
              border: 0,
              width: "100%",
              height: "100%",
            }}
            src={node.attribs?.src}
            allowFullScreen
          />
        </div>
      </Figure>
    );
  }

  if (node.name === "figcaption") {
    return (
      <figcaption
        key={index}
        css={{
          textAlign: "center",
          margin: 0,
          marginTop: "0.625rem",
          color: "var(--color-text-tertiary)",
          fontSize: "0.8125rem",
          lineHeight: 1.5,
          letterSpacing: "0.00625em",
        }}
      >
        {node.children.map(transform)}
      </figcaption>
    );
  }

  if (node.name === "img") {
    const { alt, src } = node.attribs;

    return (
      <img
        key={index}
        css={{
          width: "100%",
          borderRadius: "0.5rem",
        }}
        src={src}
        alt={alt}
        loading="lazy"
      />
    );
  }

  if (node.name === "blockquote") {
    return (
      <blockquote
        key={index}
        css={{
          margin: 0,
          padding: 0,
          color: "var(--color-highlight)",
          fontFamily: fonts.monospace,
          fontSize: "1.5rem",
          fontWeight: 400,
          marginTop: "1rem",
          marginBottom: "2rem",
          lineHeight: 1.5,

          [media(breakpoints.medium)]: {
            padding: "2rem 3rem",
          },
        }}
      >
        {node.children.map(transform)}
      </blockquote>
    );
  }

  if (node.name === "table") {
    return <Table key={index}>{node.children.map(transform)}</Table>;
  }

  if (node.type === "text") {
    return <KatexText text={node.data} key={index} />;
  }

  const element = convertNodeToElement(node, index, transform);

  return element;
};

export interface RichTextProps {
  html: string;
}

/**
 * XSS-safe HTML renderer.
 *
 * @param {React.PropsWithChildren<RichTextProps>} props Props.
 *
 * @returns {React.ReactElement} The rendered HTML.
 */
const RichText: FunctionComponent<RichTextProps> = ({ html, ...props }) => {
  const parsedHtml = useMemo(() => {
    if (html) {
      return ReactHtmlParser(html, { transform });
    }

    return (
      <>
        <h2>
          <InlineSkeleton />
        </h2>
        <p>
          <InlineSkeleton count={5} />
        </p>
        <Skeleton height="20rem" />
      </>
    );
  }, [html]);

  return (
    <div
      css={{
        p: {
          lineHeight: 2,
          letterSpacing: "0.00625em",
          overflowX: "hidden",
        },

        "> *": {
          marginTop: "1rem",
          marginBottom: "2rem",
        },

        h2: {
          fontSize: "1.75rem",
          lineHeight: 1.2,
          fontWeight: 700,
          letterSpacing: "-0.025em",

          [media(breakpoints.medium)]: {
            fontSize: "2rem",
          },

          [media(breakpoints.large)]: {
            fontSize: "2.25rem",
          },
        },

        h3: {
          fontSize: "1.375rem",
          lineHeight: 1.5,

          [media(breakpoints.medium)]: {
            fontSize: "1.5rem",
          },
        },

        li: {
          marginBottom: "0.5rem",
          lineHeight: 1.7,
        },

        hr: {
          border: 0,
          borderTop: "1px solid var(--color-border-primary)",
        },
      }}
      {...props}
    >
      {parsedHtml}
    </div>
  );
};

export default RichText;
