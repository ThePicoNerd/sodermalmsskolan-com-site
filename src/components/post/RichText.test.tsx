import { render } from "@testing-library/react";
import React from "react";
import RichText from "./RichText";

const html = `
<h1>Title</h1>
<h2>Smaller title</h2>
<script>
alert("oof");
</script>
<a href="https://www.youtube.com/watch?v=dQw4w9WgXcQ">Cool video</a>
<figure>
  <img src="https://media1.tenor.com/images/467d353f7e2d43563ce13fddbb213709/tenor.gif?itemid=12136175" />
  <figcaption>my favorite gif</figcaption>
</figure>
<table>
  <thead>
    <tr>
      <th>Company</th>
      <th>Contact</th>
      <th>Country</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Alfreds Futterkiste</td>
      <td>Maria Anders</td>
      <td>Germany</td>
    </tr>
    <tr>
      <td>Centro comercial Moctezuma</td>
      <td>Francisco Chang</td>
      <td>Mexico</td>
    </tr>
  </tbody>
</table>
<iframe width="560" height="315" src="https://www.youtube.com/embed/IvUU8joBb1Q" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
<blockquote>deep quote</blockquote>
`;

describe("<RichText /> test", () => {
  it("renders as usual", () => {
    const result = render(<RichText html={html} />);

    expect(result.asFragment()).toMatchSnapshot();
  });

  it("returns a skeleton if there is no HTML", () => {
    const result = render(<RichText html="" />);

    expect(result.asFragment()).toMatchSnapshot();
  });
});
