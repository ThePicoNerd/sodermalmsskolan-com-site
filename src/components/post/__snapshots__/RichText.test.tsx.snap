// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RichText /> test renders as usual 1`] = `
<DocumentFragment>
  .emotion-0 p {
  line-height: 2;
  letter-spacing: 0.00625em;
  overflow-x: hidden;
}

.emotion-0>* {
  margin-top: 1rem;
  margin-bottom: 2rem;
}

.emotion-0 h2 {
  font-size: 1.75rem;
  line-height: 1.2;
  font-weight: 700;
  letter-spacing: -0.025em;
}

@media screen and (min-width: 768px) {
  .emotion-0 h2 {
    font-size: 2rem;
  }
}

@media screen and (min-width: 1024px) {
  .emotion-0 h2 {
    font-size: 2.25rem;
  }
}

.emotion-0 h3 {
  font-size: 1.375rem;
  line-height: 1.5;
}

@media screen and (min-width: 768px) {
  .emotion-0 h3 {
    font-size: 1.5rem;
  }
}

.emotion-0 li {
  margin-bottom: 0.5rem;
  line-height: 1.7;
}

.emotion-0 hr {
  border: 0;
  border-top: 1px solid var(--color-border-primary);
}

.emotion-1 {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-1:hover {
  color: var(--link-color);
}

.emotion-2 {
  --figure-margin-x: 0px;
  --figure-margin-y: 4rem;
  margin: var(--figure-margin-y) var(--figure-margin-x);
}

@media screen and (min-width: 480px) {
  .emotion-2 {
    --figure-margin-x: -2rem;
  }
}

@media screen and (min-width: 768px) {
  .emotion-2 {
    --figure-margin-x: -3rem;
  }
}

.emotion-3 {
  width: 100%;
  border-radius: 0.5rem;
}

.emotion-4 {
  text-align: center;
  margin: 0;
  margin-top: 0.625rem;
  color: var(--color-text-tertiary);
  font-size: 0.8125rem;
  line-height: 1.5;
  letter-spacing: 0.00625em;
}

.emotion-5 {
  overflow-x: auto;
  -webkit-overflow-scrolling: touch;
}

.emotion-6 {
  border: 1px solid var(--color-border-primary);
  border-spacing: 0;
  border-radius: 0.5rem;
  overflow: hidden;
}

.emotion-6 tr:not(:last-child) td,
.emotion-6 thead th {
  border-bottom: 1px solid var(--color-border-primary);
}

.emotion-6 tbody tr {
  -webkit-transition: background-color 0.2s ease;
  transition: background-color 0.2s ease;
}

.emotion-6 tbody tr:nth-of-type(2n + 1) {
  background: var(--color-bg-secondary);
}

.emotion-6 tbody tr:hover {
  background: var(--color-border-primary);
}

.emotion-6 td,
.emotion-6 th {
  padding: 0.75rem;
  text-align: left;
}

.emotion-6 td:not(:last-child),
.emotion-6 th:not(:last-child) {
  border-right: 1px solid var(--color-border-primary);
}

.emotion-6 th {
  font-weight: 700;
}

.emotion-8 {
  position: relative;
  display: block;
  width: 100%;
  overflow: hidden;
}

.emotion-8::before {
  padding-top: 56.25rem;
  display: block;
  content: "";
}

.emotion-9 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border: 0;
  width: 100%;
  height: 100%;
}

.emotion-10 {
  margin: 0;
  padding: 0;
  color: var(--color-highlight);
  font-family: IBM Plex Mono,monospace;
  font-size: 1.5rem;
  font-weight: 400;
  margin-top: 1rem;
  margin-bottom: 2rem;
  line-height: 1.5;
}

@media screen and (min-width: 768px) {
  .emotion-10 {
    padding: 2rem 3rem;
  }
}

<div
    class="emotion-0"
  >
    <h1>
      Title
    </h1>
    <h2>
      Smaller title
    </h2>
    <p>
      
  How does it handle
  multiple
  lines?

    </p>
    <a
      class="emotion-1"
      href="https://www.youtube.com/watch?v=dQw4w9WgXcQ"
    >
      Cool video
    </a>
    <div
      class="emotion-2"
    >
      <img
        class="emotion-3"
        loading="lazy"
        src="https://media1.tenor.com/images/467d353f7e2d43563ce13fddbb213709/tenor.gif?itemid=12136175"
      />
      <figcaption
        class="emotion-4"
      >
        my favorite gif
      </figcaption>
    </div>
    <div
      class="emotion-5"
    >
      <table
        class="emotion-6"
      >
        <thead>
          <tr>
            <th>
              Company
            </th>
            <th>
              Contact
            </th>
            <th>
              Country
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Alfreds Futterkiste
            </td>
            <td>
              Maria Anders
            </td>
            <td>
              Germany
            </td>
          </tr>
          <tr>
            <td>
              Centro comercial Moctezuma
            </td>
            <td>
              Francisco Chang
            </td>
            <td>
              Mexico
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="emotion-2"
    >
      <div
        class="emotion-8"
      >
        <iframe
          allowfullscreen=""
          class="emotion-9"
          src="https://www.youtube.com/embed/IvUU8joBb1Q"
        />
      </div>
    </div>
    <blockquote
      class="emotion-10"
    >
      deep quote
    </blockquote>
  </div>
</DocumentFragment>
`;

exports[`<RichText /> test returns a skeleton if there is no HTML 1`] = `
<DocumentFragment>
  @keyframes animation-0 {
  0% {
    -webkit-background-position: 200% 0;
    background-position: 200% 0;
  }

  100% {
    -webkit-background-position: -200% 0;
    background-position: -200% 0;
  }
}

@keyframes animation-0 {
  0% {
    -webkit-background-position: 200% 0;
    background-position: 200% 0;
  }

  100% {
    -webkit-background-position: -200% 0;
    background-position: -200% 0;
  }
}

.emotion-0 p {
  line-height: 2;
  letter-spacing: 0.00625em;
  overflow-x: hidden;
}

.emotion-0>* {
  margin-top: 1rem;
  margin-bottom: 2rem;
}

.emotion-0 h2 {
  font-size: 1.75rem;
  line-height: 1.2;
  font-weight: 700;
  letter-spacing: -0.025em;
}

@media screen and (min-width: 768px) {
  .emotion-0 h2 {
    font-size: 2rem;
  }
}

@media screen and (min-width: 1024px) {
  .emotion-0 h2 {
    font-size: 2.25rem;
  }
}

.emotion-0 h3 {
  font-size: 1.375rem;
  line-height: 1.5;
}

@media screen and (min-width: 768px) {
  .emotion-0 h3 {
    font-size: 1.5rem;
  }
}

.emotion-0 li {
  margin-bottom: 0.5rem;
  line-height: 1.7;
}

.emotion-0 hr {
  border: 0;
  border-top: 1px solid var(--color-border-primary);
}

.emotion-1 {
  background-color: var(--skeleton-base);
  background-image: linear-gradient(270deg, var(--skeleton-base), var(--skeleton-highlight), var(--skeleton-base));
  background-repeat: no-repeat;
  -webkit-background-size: 200% 100%;
  background-size: 200% 100%;
  -webkit-animation: animation-0 2s ease-in-out infinite;
  animation: animation-0 2s ease-in-out infinite;
  border-radius: 0.3125rem;
  display: inline-block;
  line-height: 1;
  width: 100%;
}

.emotion-7 {
  background-color: var(--skeleton-base);
  background-image: linear-gradient(270deg, var(--skeleton-base), var(--skeleton-highlight), var(--skeleton-base));
  background-repeat: no-repeat;
  -webkit-background-size: 200% 100%;
  background-size: 200% 100%;
  -webkit-animation: animation-0 2s ease-in-out infinite;
  animation: animation-0 2s ease-in-out infinite;
  border-radius: 0.3125rem;
  display: inline-block;
  line-height: 1;
  width: 100%;
  height: 20rem;
}

<div
    class="emotion-0"
  >
    <h2>
      <span>
        <span
          class="emotion-1"
        >
          ‌
        </span>
      </span>
    </h2>
    <p>
      <span>
        <span
          class="emotion-1"
        >
          ‌
        </span>
        <span
          class="emotion-1"
        >
          ‌
        </span>
        <span
          class="emotion-1"
        >
          ‌
        </span>
        <span
          class="emotion-1"
        >
          ‌
        </span>
        <span
          class="emotion-1"
        >
          ‌
        </span>
      </span>
    </p>
    <span
      class="emotion-7"
    />
  </div>
</DocumentFragment>
`;
